File: app/scripts/sheet.py
import json,subprocess,sys
from PIL import Image, ImageDraw
base,out,pieces = sys.argv[1],sys.argv[2],sys.argv[3]
angles = sys.argv[4] if len(sys.argv)>4 else '35:18,140:16,-55:10'
rov = sys.argv[5] if len(sys.argv)>5 else ''
r=subprocess.run(['node','scripts/sheet.mjs',base,out,pieces,angles,rov],capture_output=True,text=True)
line=[l for l in r.stdout.strip().split('\n') if l.startswith('{')][-1]
d=json.loads(line); cols=d['cols']; shots=d['shots']
rows=(len(shots)+cols-1)//cols
W,H=560,480
im=Image.new('RGB',(W*cols,H*rows),(18,14,9)); dr=ImageDraw.Draw(im)
for i,(piece,a,e,f,errs) in enumerate(shots):
    im.paste(Image.open(f),(W*(i%cols),H*(i//cols)))
    dr.text((W*(i%cols)+8,H*(i//cols)+6),f"{piece}  a{a} e{e}"+(f"  ERR{errs}" if errs else ""),fill=(255,220,150))
im.save(out); print(out, im.size)

File: app/scripts/sheet.mjs
/**
 * Contact sheet: render several pieces at several angles into one image grid.
 *   node scripts/sheet.mjs <base> <out.png> <piece>[,piece...] [angles]
 */
import { chromium } from 'playwright'
const base = process.argv[2], out = process.argv[3]
const pieces = process.argv[4].split(',')
const angles = (process.argv[5] ?? '35:18,140:16,-55:10').split(',').map(s => s.split(':'))
const rOv = process.argv[6] ?? ''
const b = await chromium.launch({
  executablePath: '/opt/pw-browsers/chromium',
  args: ['--use-gl=swiftshader', '--enable-unsafe-swiftshader', '--no-sandbox', '--disable-dev-shm-usage'],
})
const shots = []
for (const piece of pieces) {
  for (const [a, e] of angles) {
    const p = await b.newPage({ viewport: { width: 560, height: 480 } })
    const errs = []
    p.on('pageerror', x => errs.push(x.message))
    await p.goto(`${base}?debug=piece&p=${piece}&a=${a}&e=${e}${rOv ? `&r=${rOv}` : ''}`, { waitUntil: 'domcontentloaded', timeout: 60000 })
    await p.waitForTimeout(5000)
    const f = `/tmp/_sheet_${piece}_${a}_${e}.png`
    await p.screenshot({ path: f, animations: 'disabled', timeout: 120000 })
    shots.push([piece, a, e, f, errs.length])
    await p.close()
  }
}
await b.close()
console.log(JSON.stringify({ out, cols: angles.length, shots }))
